pkg/commands/artifact/run.go

```diff
@@ -461,20 +461,34 @@ func initScannerConfig(opts flag.Options, cacheClient cache.Cache) (ScannerConfi
 		log.Logger.Debugf("Vulnerability type:  %s", scanOptions.VulnType)
 	}
 
+	var downloadedPolicyPaths []string
+	var disableEmbedded bool
+	if len(opts.PolicyPaths) == 0 {
```

Comment:
Yes this is correct. There is a good reason for this however and that is the fact that if we do load embedded policies when a user specifies a custom policy of their own, we risk the chance of getting rego namespace collisions. 

A more detailed explanation with an example: A user loads a custom policy that they've written that uses rego helper functions from an aqua library, like this: https://github.com/aquasecurity/defsec/blob/master/internal/rules/policies/kubernetes/lib/kubernetes.rego - in this case there's a chance that their code is using a version of the helper library (kubernetes.rego) that is out of date or different from the latest embedded helper library (kubernetes.rego). This would cause their custom policy to fail.

So in short, if a user loads a custom policy, it is on them to supply all necessary helper libraries as well for the policy to work properly. This eliminates any confusion or mismatch as described above.

Reply:
It makes sense that we don't want to load helper functions when custom policies are passed.
But if I understand correctly, the current version loads custom policies in addition to built-in policies, right? Will we change the behavior to disable all built-in policies when custom policies are specified? Isn't this a breaking change?

Reply:
Hm yeah that is a good point. Although the breakage is limited to those custom policies that use any Rego helpers we provide. 

Do you have any proposal on how we can if possible, avoid this breaking change?

Reply:
It is inconvenient if the built-in policies are disabled and not intuitive as users assume custom policies are add-on. I think we can accept the risk of out-of-date helper functions. We should warn it in our doc though. 
@simar7 Did you already see any issue from the version mismatch of the helper library?
@itaysk Any thoughts?

Reply:
Thanks for catching this @knqyf263 I agree, I don't think it's reasonable to expect the user to provide helpers for Trivy's built-in policies. Agree with your recommendation.

Reply:
Changing the namespace of our helpers sounds good to me. It could be a breaking change, so we can announce that along with the embedded policy retirement.

Reply:
Actually I might have missed a thing here as it's been a while I made this PR. So I tested out the current set of changes and it is in fact loading both embedded policies and custom policies (if any are specified). 

IOW loading custom policies does not disable embedded policies (current behaviour is maintained).

Also if a custom policy uses a helper, it will use the one that the embedding policies have (also current behaviour).

See below for an example. Note there are 4 results as the "custom" policy I'm loading is just a copy of an embedded policy, DS005.

```shell
rm -rf  /Users/simar/Library/Caches/trivy/policy  &&  ./trivy --debug conf --skip-policy-update /Users/simar/repos/defsec/test/testdata/dockerfile/DS005 --config-policy=/Users/simar/repos/defsec-policies-testing



2022-12-12T13:22:47.745-0800    DEBUG   Severities: ["UNKNOWN" "LOW" "MEDIUM" "HIGH" "CRITICAL"]
2022-12-12T13:22:47.800-0800    DEBUG   cache dir:  /Users/simar/Library/Caches/trivy
>>> policy paths [/Users/simar/repos/defsec-policies-testing]
2022-12-12T13:22:47.800-0800    INFO    Misconfiguration scanning is enabled
2022-12-12T13:22:47.800-0800    DEBUG   Walk the file tree rooted at '/Users/simar/repos/defsec/test/testdata/dockerfile/DS005' in parallel
2022-12-12T13:22:48.254-0800    DEBUG   OS is not detected.
2022-12-12T13:22:48.254-0800    INFO    Detected config files: 2
2022-12-12T13:22:48.254-0800    DEBUG   Scanned config file: Dockerfile.allowed
2022-12-12T13:22:48.254-0800    DEBUG   Scanned config file: Dockerfile.denied

Dockerfile.denied (dockerfile)

Tests: 25 (SUCCESSES: 21, FAILURES: 4, EXCEPTIONS: 0)
Failures: 4 (UNKNOWN: 0, LOW: 4, MEDIUM: 0, HIGH: 0, CRITICAL: 0)

LOW: CUSTOM Consider using 'COPY "/target/app.jar" "app.jar"' command instead of 'ADD "/target/app.jar" "app.jar"'
═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
You should use COPY instead of ADD unless you want to extract a tar file. Note that an ADD command will extract a tar file, which adds the risk of Zip-based vulnerabilities. Accordingly, it is advised to use a COPY command, which does not extract tar files.

See https://avd.aquasec.com/misconfig/ds005
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 Dockerfile.denied:4
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   4 [ ADD "/target/app.jar" "app.jar"
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────


LOW: CUSTOM Consider using 'COPY "/target/app.jar" "app.jar"' command instead of 'ADD "/target/app.jar" "app.jar"'
═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
You should use COPY instead of ADD unless you want to extract a tar file. Note that an ADD command will extract a tar file, which adds the risk of Zip-based vulnerabilities. Accordingly, it is advised to use a COPY command, which does not extract tar files.

See https://avd.aquasec.com/misconfig/ds005
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 Dockerfile.denied:4
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   4 [ ADD "/target/app.jar" "app.jar"
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────


LOW: Consider using 'COPY "/target/app.jar" "app.jar"' command instead of 'ADD "/target/app.jar" "app.jar"'
═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
You should use COPY instead of ADD unless you want to extract a tar file. Note that an ADD command will extract a tar file, which adds the risk of Zip-based vulnerabilities. Accordingly, it is advised to use a COPY command, which does not extract tar files.

See https://avd.aquasec.com/misconfig/ds005
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 Dockerfile.denied:4
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   4 [ ADD "/target/app.jar" "app.jar"
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────


LOW: Consider using 'COPY "/target/app.jar" "app.jar"' command instead of 'ADD "/target/app.jar" "app.jar"'
═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
You should use COPY instead of ADD unless you want to extract a tar file. Note that an ADD command will extract a tar file, which adds the risk of Zip-based vulnerabilities. Accordingly, it is advised to use a COPY command, which does not extract tar files.

See https://avd.aquasec.com/misconfig/ds005
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 Dockerfile.denied:4
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   4 [ ADD "/target/app.jar" "app.jar"
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

```

Policy can be found here: https://github.com/aquasecurity/defsec/blob/master/internal/rules/policies/docker/policies/add_instead_of_copy.rego

Reply:
If this is the actually behavior (not sure since Teppei spotted the behavior in the code) then it sounds like a good one to me. To summarize:

1. Built-in policies are downloaded by default
  1.1 can be disabled `--skip-policy-update`
2. Built-in policies are loaded by default
  2.1 can not be disabled (right?)
3. Custom policies, if provided `--config-policy`, are loaded in addition to built-in
3. Built-in helpers are always loaded and available for built-in and custom policies
  3.1 if built-in policies bundle was not downloaded, built-in helpers are not available for custom policy author (correct? might want to fix it in the future)
  3.2 TODO: namespace built-in helpers

Reply:
>IOW loading custom policies does not disable embedded policies (current behaviour is maintained).

Oh, seems like we don't use downloaded polices, but embedded policies. If custom policies are passed, we switch downloaded policies to embedded policies. This is why I said embedded policies complicate the logic.

For example,

- The latest downloaded policies: Policy A, B, C (C was added to defsec recently)
- The embedded policies: Policy A, B
- The user custom policy: Policy D

In this case, `trivy config /path/to/dir` detects A, B and C, but `trivy config --config-policy ./my-policy /path/to/dir` will detect A, B and D, not detect C anymore. I don't think it is expected behavior.

Reply:
so if we consider the typical hierarchy:
1. embedded
2. downloaded
3. custom

it should be loaded in this order, meaning 3 overrides 2 overrides 1

currently we miss step 2 if custom if provided.
correct?

Reply:
IIUC, this PR behaves as below:

1. Built-in policies are downloaded by default
    1.1 can be disabled --skip-policy-update
1. Built-in policies are loaded by default
    2.1 can not be disabled (can be ignored by using exception policies though)
1. Custom policies, if provided --config-policy, are loaded in addition to **embedded built-in** policies
    3.1 Regardless of whether built-in policies are downloaded
1. Built-in helpers are always loaded and available for built-in and custom policies

It should be like:

1. Built-in policies are downloaded by default
    1.1 can be disabled --skip-policy-update
1. Built-in policies are loaded by default
    2.1 can not be disabled (can be ignored by using exception policies though)
1. Custom policies, if provided --config-policy, are loaded in addition to built-in policies
    3.1 Use downloaded policies if present
    3.2 Use embedded policies if there are no downloaded policies in local
1. Built-in helpers are always loaded and available for built-in and custom policies

Reply:
1. built-in
    1.1. embedded
    1.2. downloaded
1. custom

`1.2` should override `1.1`. `2` should not override `1`.

Reply:
Thanks for the positive discussion from you both, I've read through it and I think I have an idea of what's desired.

> 1. built-in
>    1.1. embedded
>    1.2. downloaded (if present)
> 2. custom
> 
> `1.2` should override `1.1`. `2` should not override `1`. Enabling `2` should not affect `1`. `1.1` or `1.2` should be selected regardless of `2`.

Just to recap, I'm going by this as it's the most succinct and clear definition so far.


Here's a few results, please point out any logical flaws if you see any:

1. No cache, `--skip-policy-update` enabled and No custom policy passed:
```shell
rm -rf  /Users/simar/Library/Caches/trivy/policy  &&  ./trivy --skip-policy-update --debug  conf /Users/simar/repos/defsec-policies-testing
2023-01-06T19:05:33.343-0800	DEBUG	Severities: ["UNKNOWN" "LOW" "MEDIUM" "HIGH" "CRITICAL"]
2023-01-06T19:05:33.385-0800	DEBUG	cache dir:  /Users/simar/Library/Caches/trivy
2023-01-06T19:05:33.385-0800	INFO	No downloadable policies were loaded as --skip-policy-update is enabled
2023-01-06T19:05:33.386-0800	DEBUG	Falling back to embedded policies: No downloadable policies were loaded as --skip-policy-update is enabled
2023-01-06T19:05:33.386-0800	INFO	Misconfiguration scanning is enabled
2023-01-06T19:05:33.386-0800	DEBUG	Walk the file tree rooted at '/Users/simar/repos/defsec-policies-testing' in parallel
2023-01-06T19:05:33.950-0800	DEBUG	OS is not detected.
2023-01-06T19:05:33.951-0800	INFO	Detected config files: 1
2023-01-06T19:05:33.951-0800	DEBUG	Scanned config file: Dockerfile.denied

Dockerfile.denied (dockerfile)

Tests: 22 (SUCCESSES: 21, FAILURES: 1, EXCEPTIONS: 0)
Failures: 1 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 1)

CRITICAL: 'COPY --from' should not mention current alias 'dep' since it is impossible to copy from itself
═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
COPY '--from' should not mention the current FROM alias, since it is impossible to copy from itself.

See https://avd.aquasec.com/misconfig/ds006
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 Dockerfile.denied:2
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   2 [ COPY --from=dep /binary /
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```

**Result: Only embedded policy is used.**

2. No cache, `--skip-policy-update` enabled and ONE custom policy passed:
```shell
rm -rf  /Users/simar/Library/Caches/trivy/policy  &&  ./trivy --skip-policy-update --debug  conf /Users/simar/repos/defsec-policies-testing --config-policy=/Users/simar/repos/defsec-policies-testing
2023-01-06T19:04:26.421-0800	DEBUG	Severities: ["UNKNOWN" "LOW" "MEDIUM" "HIGH" "CRITICAL"]
2023-01-06T19:04:26.480-0800	DEBUG	cache dir:  /Users/simar/Library/Caches/trivy
2023-01-06T19:04:26.480-0800	INFO	No downloadable policies were loaded as --skip-policy-update is enabled
2023-01-06T19:04:26.480-0800	DEBUG	Falling back to embedded policies: No downloadable policies were loaded as --skip-policy-update is enabled
2023-01-06T19:04:26.480-0800	INFO	Misconfiguration scanning is enabled
2023-01-06T19:04:26.480-0800	DEBUG	Walk the file tree rooted at '/Users/simar/repos/defsec-policies-testing' in parallel
2023-01-06T19:04:27.007-0800	DEBUG	OS is not detected.
2023-01-06T19:04:27.007-0800	INFO	Detected config files: 1
2023-01-06T19:04:27.007-0800	DEBUG	Scanned config file: Dockerfile.denied

Dockerfile.denied (dockerfile)

Tests: 23 (SUCCESSES: 21, FAILURES: 2, EXCEPTIONS: 0)
Failures: 2 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 2)

CRITICAL: 'COPY --from' should not mention current alias 'dep' since it is impossible to copy from itself
═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
COPY '--from' should not mention the current FROM alias, since it is impossible to copy from itself.

See https://avd.aquasec.com/misconfig/ds006
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 Dockerfile.denied:2
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   2 [ COPY --from=dep /binary /
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────


CRITICAL: My amazing new custom policy
═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
CUSTOM POLICY: My amazing new custom policy.

See https://avd.aquasec.com/misconfig/ds999
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 Dockerfile.denied:2
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   2 [ COPY --from=dep /binary /
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```
**Result: Observe that custom policy was scanned in addition to the embedded policy as desired by: `2 should not override 1. Enabling 2 should not affect 1. 1.1 or 1.2 should be selected regardless of 2.`**

3. No cache, policy update is successful, No custom policy:
```shell
rm -rf  /Users/simar/Library/Caches/trivy/policy  &&  ./trivy  --debug  conf /Users/simar/repos/defsec-policies-testing
2023-01-06T19:09:15.973-0800	DEBUG	Severities: ["UNKNOWN" "LOW" "MEDIUM" "HIGH" "CRITICAL"]
2023-01-06T19:09:16.027-0800	DEBUG	cache dir:  /Users/simar/Library/Caches/trivy
2023-01-06T19:09:16.027-0800	DEBUG	Failed to open the policy metadata: open /Users/simar/Library/Caches/trivy/policy/metadata.json: no such file or directory
2023-01-06T19:09:16.027-0800	INFO	Need to update the built-in policies
2023-01-06T19:09:16.027-0800	INFO	Downloading the built-in policies...
30.72 KiB / 30.72 KiB [------------------------------------------------------------------------------------] 100.00% ? p/s 0s
2023-01-06T19:09:16.208-0800	DEBUG	Digest of the built-in policies: sha256:e089eac12028c59cdbdc9ae21653ab5e0600d146fe5cc1fad1535f3446118802
2023-01-06T19:09:16.209-0800	DEBUG	Policies successfully loaded from disk
2023-01-06T19:09:16.209-0800	INFO	Misconfiguration scanning is enabled
2023-01-06T19:09:16.209-0800	DEBUG	Walk the file tree rooted at '/Users/simar/repos/defsec-policies-testing' in parallel
2023-01-06T19:09:16.757-0800	DEBUG	OS is not detected.
2023-01-06T19:09:16.757-0800	INFO	Detected config files: 1
2023-01-06T19:09:16.757-0800	DEBUG	Scanned config file: Dockerfile.denied

Dockerfile.denied (dockerfile)

Tests: 22 (SUCCESSES: 21, FAILURES: 1, EXCEPTIONS: 0)
Failures: 1 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 1)

CRITICAL: 'COPY --from' should not mention current alias 'dep' since it is impossible to copy from itself
═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
DOWNLOADED POLICY: COPY '--from' should not mention the current FROM alias, since it is impossible to copy from itself.

See https://avd.aquasec.com/misconfig/ds006
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 Dockerfile.denied:2
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   2 [ COPY --from=dep /binary /
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```

**Result: Only newer downloaded policy is used.**

4. No cache, policy update is successful, ONE custom policy:
```shell
rm -rf  /Users/simar/Library/Caches/trivy/policy  &&  ./trivy --debug  conf /Users/simar/repos/defsec-policies-testing --config-policy=/Users/simar/repos/defsec-policies-testing
2023-01-06T19:10:45.034-0800	DEBUG	Severities: ["UNKNOWN" "LOW" "MEDIUM" "HIGH" "CRITICAL"]
2023-01-06T19:10:45.088-0800	DEBUG	cache dir:  /Users/simar/Library/Caches/trivy
2023-01-06T19:10:45.088-0800	DEBUG	Failed to open the policy metadata: open /Users/simar/Library/Caches/trivy/policy/metadata.json: no such file or directory
2023-01-06T19:10:45.088-0800	INFO	Need to update the built-in policies
2023-01-06T19:10:45.088-0800	INFO	Downloading the built-in policies...
30.72 KiB / 30.72 KiB [------------------------------------------------------------------------------------] 100.00% ? p/s 0s
2023-01-06T19:10:45.263-0800	DEBUG	Digest of the built-in policies: sha256:e089eac12028c59cdbdc9ae21653ab5e0600d146fe5cc1fad1535f3446118802
2023-01-06T19:10:45.263-0800	DEBUG	Policies successfully loaded from disk
2023-01-06T19:10:45.263-0800	INFO	Misconfiguration scanning is enabled
2023-01-06T19:10:45.264-0800	DEBUG	Walk the file tree rooted at '/Users/simar/repos/defsec-policies-testing' in parallel
2023-01-06T19:10:45.867-0800	DEBUG	OS is not detected.
2023-01-06T19:10:45.867-0800	INFO	Detected config files: 1
2023-01-06T19:10:45.867-0800	DEBUG	Scanned config file: Dockerfile.denied

Dockerfile.denied (dockerfile)

Tests: 23 (SUCCESSES: 21, FAILURES: 2, EXCEPTIONS: 0)
Failures: 2 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 2)

CRITICAL: 'COPY --from' should not mention current alias 'dep' since it is impossible to copy from itself
═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
DOWNLOADED POLICY: COPY '--from' should not mention the current FROM alias, since it is impossible to copy from itself.

See https://avd.aquasec.com/misconfig/ds006
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 Dockerfile.denied:2
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   2 [ COPY --from=dep /binary /
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────


CRITICAL: My amazing new custom policy
═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
CUSTOM POLICY: My amazing new custom policy.

See https://avd.aquasec.com/misconfig/ds999
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 Dockerfile.denied:2
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   2 [ COPY --from=dep /binary /
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```
**Result: Observe that again Custom policy is scanned in addition but this time it is the NEWER downloaded policy. This is as per: `1.2 should override 1.1`.**

5. No cache, policy update fails, NO custom policy
```shell
rm -rf  /Users/simar/Library/Caches/trivy/policy  &&  ./trivy --debug  conf /Users/simar/repos/defsec-policies-testing
2023-01-06T19:12:51.734-0800	DEBUG	Severities: ["UNKNOWN" "LOW" "MEDIUM" "HIGH" "CRITICAL"]
2023-01-06T19:12:51.790-0800	DEBUG	cache dir:  /Users/simar/Library/Caches/trivy
2023-01-06T19:12:51.790-0800	DEBUG	Failed to open the policy metadata: open /Users/simar/Library/Caches/trivy/policy/metadata.json: no such file or directory
2023-01-06T19:12:51.790-0800	INFO	Need to update the built-in policies
2023-01-06T19:12:51.790-0800	INFO	Downloading the built-in policies...
2023-01-06T19:12:51.931-0800	DEBUG	Falling back to embedded policies: failed to download built-in policies: OPA bundle error: OCI artifact error: OCI repository error: Get "https://localhost:5000/v2/": dial tcp [::1]:5000: connect: connection refused; Get "http://localhost:5000/v2/": dial tcp [::1]:5000: connect: connection refused
2023-01-06T19:12:51.931-0800	INFO	Misconfiguration scanning is enabled
2023-01-06T19:12:51.932-0800	DEBUG	Walk the file tree rooted at '/Users/simar/repos/defsec-policies-testing' in parallel
2023-01-06T19:12:52.445-0800	DEBUG	OS is not detected.
2023-01-06T19:12:52.445-0800	INFO	Detected config files: 1
2023-01-06T19:12:52.445-0800	DEBUG	Scanned config file: Dockerfile.denied

Dockerfile.denied (dockerfile)

Tests: 22 (SUCCESSES: 21, FAILURES: 1, EXCEPTIONS: 0)
Failures: 1 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 1)

CRITICAL: 'COPY --from' should not mention current alias 'dep' since it is impossible to copy from itself
═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
COPY '--from' should not mention the current FROM alias, since it is impossible to copy from itself.

See https://avd.aquasec.com/misconfig/ds006
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 Dockerfile.denied:2
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   2 [ COPY --from=dep /binary /
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```

**Result: Observe in this case the embedded policy is used as no new policies were downloaded.**

7. No cache, policy update fails, ONE custom policy:
```shell
rm -rf  /Users/simar/Library/Caches/trivy/policy  &&  ./trivy --debug  conf /Users/simar/repos/defsec-policies-testing --config-policy=/Users/simar/repos/defsec-policies-testing
2023-01-06T19:11:57.546-0800	DEBUG	Severities: ["UNKNOWN" "LOW" "MEDIUM" "HIGH" "CRITICAL"]
2023-01-06T19:11:57.602-0800	DEBUG	cache dir:  /Users/simar/Library/Caches/trivy
2023-01-06T19:11:57.602-0800	DEBUG	Failed to open the policy metadata: open /Users/simar/Library/Caches/trivy/policy/metadata.json: no such file or directory
2023-01-06T19:11:57.602-0800	INFO	Need to update the built-in policies
2023-01-06T19:11:57.602-0800	INFO	Downloading the built-in policies...
2023-01-06T19:11:57.738-0800	DEBUG	Falling back to embedded policies: failed to download built-in policies: OPA bundle error: OCI artifact error: OCI repository error: Get "https://localhost:5000/v2/": dial tcp [::1]:5000: connect: connection refused; Get "http://localhost:5000/v2/": dial tcp [::1]:5000: connect: connection refused
2023-01-06T19:11:57.738-0800	INFO	Misconfiguration scanning is enabled
2023-01-06T19:11:57.738-0800	DEBUG	Walk the file tree rooted at '/Users/simar/repos/defsec-policies-testing' in parallel
2023-01-06T19:11:58.306-0800	DEBUG	OS is not detected.
2023-01-06T19:11:58.306-0800	INFO	Detected config files: 1
2023-01-06T19:11:58.306-0800	DEBUG	Scanned config file: Dockerfile.denied

Dockerfile.denied (dockerfile)

Tests: 23 (SUCCESSES: 21, FAILURES: 2, EXCEPTIONS: 0)
Failures: 2 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 2)

CRITICAL: 'COPY --from' should not mention current alias 'dep' since it is impossible to copy from itself
═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
COPY '--from' should not mention the current FROM alias, since it is impossible to copy from itself.

See https://avd.aquasec.com/misconfig/ds006
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 Dockerfile.denied:2
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   2 [ COPY --from=dep /binary /
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────


CRITICAL: My amazing new custom policy
═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
CUSTOM POLICY: My amazing new custom policy.

See https://avd.aquasec.com/misconfig/ds999
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
 Dockerfile.denied:2
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   2 [ COPY --from=dep /binary /
─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```

**Result: Observe again in this case embedded policy is used (in addition to custom policy).**

Reply:
The changes to implement above are minimal but I want to confirm if the expectation is as desired before proceeding.

Reply:
Thanks for sorting out. It looks perfect to me! There is one note about no.5. We should show the fallback message as an error so that users can be aware embedded policies are used.
If there is no embedded policy, the logic can be much simplified, though. We may want to announce the retirement of embedded policies shortly.